src/common/lib/client/rest.ts

```diff
@@ -228,6 +230,155 @@ class Rest {
+  batchPublish(
+    specs: API.Types.BatchPublishSpec[],
+    callback: API.Types.StandardCallback<
+      API.Types.BatchResult<API.Types.BatchPublishSuccessResult | API.Types.BatchPublishFailureResult>[]
+    >
+  ): void;
+  batchPublish(
+    specs: API.Types.BatchPublishSpec
+  ): Promise<API.Types.BatchResult<API.Types.BatchPublishSuccessResult | API.Types.BatchPublishFailureResult>>;
+  batchPublish(
+    specs: API.Types.BatchPublishSpec[]
+  ): Promise<API.Types.BatchResult<API.Types.BatchPublishSuccessResult | API.Types.BatchPublishFailureResult>[]>;
+  batchPublish(
+    specOrSpecs: API.Types.BatchPublishSpec | API.Types.BatchPublishSpec[],
+    callbackArg?:
+      | API.Types.StandardCallback<
+          API.Types.BatchResult<API.Types.BatchPublishSuccessResult | API.Types.BatchPublishFailureResult>
+        >
+      | API.Types.StandardCallback<
+          API.Types.BatchResult<API.Types.BatchPublishSuccessResult | API.Types.BatchPublishFailureResult>[]
+        >
+  ):
+    | void
+    | Promise<API.Types.BatchResult<API.Types.BatchPublishSuccessResult | API.Types.BatchPublishFailureResult>>
+    | Promise<API.Types.BatchResult<API.Types.BatchPublishSuccessResult | API.Types.BatchPublishFailureResult>[]> {
+    if (callbackArg === undefined) {
+      if (this.options.promises) {
+        return Utils.promisify(this, 'batchPublish', [specOrSpecs]) as
+          | Promise<API.Types.BatchResult<API.Types.BatchPublishSuccessResult | API.Types.BatchPublishFailureResult>>
+          | Promise<API.Types.BatchResult<API.Types.BatchPublishSuccessResult | API.Types.BatchPublishFailureResult>[]>;
+      }
+      callbackArg = noop;
+    }
+
+    const callback = callbackArg;
+
+    let requestBodyDTO: API.Types.BatchPublishSpec[];
+    let singleSpecMode: boolean;
+    if (Array.isArray(specOrSpecs)) {
+      requestBodyDTO = specOrSpecs;
+      singleSpecMode = false;
+    } else {
+      requestBodyDTO = [specOrSpecs];
+      singleSpecMode = true;
+    }
+
+    const format = this.options.useBinaryProtocol ? Utils.Format.msgpack : Utils.Format.json,
+      headers = Utils.defaultPostHeaders(this.options, format);
+
+    if (this.options.headers) Utils.mixin(headers, this.options.headers);
+
+    const requestBody = Utils.encodeBody(requestBodyDTO, format);
+    Resource.post(
+      this,
+      '/messages',
+      requestBody,
+      headers,
+      { newBatchResponse: 'true' },
+      null,
+      (err, body, headers, unpacked) => {
+        if (err) {
+          // TODO remove this type assertion after fixing https://github.com/ably/ably-js/issues/1405
+          callback(err as API.Types.ErrorInfo);
+          return;
+        }
+
+        const batchResults = (unpacked ? body : Utils.decodeBody(body, format)) as API.Types.BatchResult<
+          API.Types.BatchPublishSuccessResult | API.Types.BatchPublishFailureResult
+        >[];
+
+        // I don't love the below type assertions for `callback` but not sure how to avoid them
```

Comment:
when using the above generic suggestion:
```typescript
const callback = callbackArg as StandardCallback<T extends BatchPublishSpec ? BatchResult<BatchPublishSuccess | BatchPublishFailure> : BatchResult<BatchPublishSuccess | BatchPublishFailure>[]>;
```

Reply:
ah yeah you're right, that wouldn't help much. fwiw i think it's possible to fix this using a typeguard but in a way that's no better than a type assertion anyway